File: 2022/july/25-07-2022/find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:

        N = len(nums)

        def findStarting():
            lo, hi = 0, N - 1
            while lo <= hi:
                mid = (lo + hi) // 2
                if nums[mid] >= target:
                    hi = mid - 1
                else:
                    lo = mid + 1
            return lo if lo < len(nums) and nums[lo] == target else -1

        def findEnding():
            lo, hi = 0, N - 1
            while lo <= hi:
                mid = (lo + hi) // 2
                if nums[mid] <= target:
                    lo = mid + 1
                else:
                    hi = mid - 1
            return hi if nums[hi] == target else -1

        return [findStarting(), findEnding()] if N > 0 else [-1, -1]

File: 2022/july/25-07-2022/find-first-and-last-position-of-element-in-sorted-array.rs
impl Solution {
    pub fn search_range(nums: Vec<i32>, target: i32) -> Vec<i32> {
        fn binary_search(nums: Vec<i32>, target: i32) -> i32{
            let (mut l, mut r) = (0, nums.len());
            while l < r{
                let mid = l + (r - l) / 2;
                if nums[mid]< target{
                    l = mid + 1;
                } else {
                    r = mid
                }
            }
            return l as i32;
        }
       let l = binary_search(nums.clone(), target);
       let r = binary_search(nums.clone(), target+1)-1; // next number of target - 1
       if l <= r {vec![l, r]} else {vec![-1, -1]}
    }
}